core/src/main/java/org/elasticsearch/env/Environment.java

```diff
@@ -332,4 +339,26 @@ public Path tmpFile() {
     public static FileStore getFileStore(Path path) throws IOException {
         return ESFileStore.getMatchingFileStore(path, fileStores);
     }
+
+    /**
+     * assserts that the two environments are equivalent for all things the environment cares about (i.e., all but the setting
+     * object which may contain different setting)
+     */
+    public static void assertEquivalent(Environment actual, Environment expected) {
```

Comment:
The idea is to indeed remove the setting field from environment and stop using it as a vehicle for settings which  are also available via other places (Node) and may or may not be the same. If we do that, then I don't need to reinitialize the environment and can use a single one, which make the assertion not needed and it will go away - no need for `equal`.

That said, the change to environment will require some other changes which may be resisted - so no promises :)